query/worker.go

```diff
@@ -0,0 +1,194 @@
+// goroutine.
+func (w *worker) run() {
+	// Subscribe to messages from the peer.
+	msgChan, cancel := w.peer.SubscribeRecvMsg()
+	defer cancel()
+
+	for {
+		var job *queryTask
+
+		select {
+		// Poll a new job from the nextJob channel.
+		case job = <-w.nextJob:
+
+		// Ignore any message received while not working on a task.
+		case <-msgChan:
+			continue
+
+		// If the peer disconnected, we can exit immediately, as we
+		// weren't working on a query.
+		case <-w.peer.OnDisconnect():
+			log.Debugf("Peer %v for worker disconnected",
+				w.peer.Addr())
+			return
+
+		case <-w.quit:
+			return
+		}
+
+		// We received a query job, send it to the peer.
+		w.peer.QueueMessageWithEncoding(
+			job.Req, nil, job.options.encoding,
+		)
+
+		// Wait for the correct response to be received from the peer,
+		// or an error happening.
+		var (
+			resp    wire.Message
+			jobErr  error
+			timeout = time.After(job.options.timeout)
+		)
+
+	Loop:
+		for {
+
+			select {
+			// A message was received from the peer, use the
+			// response handler to check whether it was answering
+			// our request.
+			case resp = <-msgChan:
+				done, progress := job.HandleResp(
+					job.Req, resp, w.peer.Addr(),
+				)
+
+				// If the response did not answer our query, we
+				// check whether it did progress it.
+				if !done {
+					// If it did make progress we reset the
+					// timeout. This ensures that the
+					// queries with multiple responses
+					// expected won't timeout before all
+					// responses have been handled.
+					// TODO(halseth): separete progress
+					// timeout value.
+					if progress {
+						timeout = time.After(
+							job.options.timeout,
+						)
+					}
+					continue Loop
+				}
+
+				// We did get a valid response, and can break
+				// the loop.
+				break Loop
+
+			// If the timeout is reached before a valid response
+			// has been received, we exit with an error.
+			case <-timeout:
+				// The query did experience a timeout and will
+				// be given to someone else.
+				jobErr = ErrQueryTimeout
+				break Loop
+
+			// If the peer disconnectes before giving us a valid
+			// answer, we'll also exit with an error.
+			case <-w.peer.OnDisconnect():
+				log.Debugf("Peer %v for worker disconnected, "+
+					"cancelling task", w.peer.Addr())
+
+				jobErr = ErrPeerDisconnected
+				break Loop
+
+			case <-w.quit:
+				return
+			}
+		}
+
+		// We have a result ready for the query task, hand it off
+		// before getting a new task.
+		select {
+		case w.results <- &jobResult{
+			task: job,
+			err:  jobErr,
+		}:
+		case <-w.quit:
+			return
+		}
+
+		switch jobErr {
+
+		// If the worker experienced a timeout trying to query this
+		// peer, we wait here for a given cooldown period before
+		// fetching the next job. This is to give other peers the
+		// chance to pick up the jobs.
+		case ErrQueryTimeout:
+			select {
+			case <-time.After(job.options.peerCooldown):
```

Comment:
Perhaps we should _randomize_ the cooldown to avoid a state where we're underutilizing the network bandwidth since all peers are paused.

Reply:
Cooldown is no more